core/chains/evm/log/broadcaster.go

```diff
@@ -350,23 +350,29 @@ func (b *broadcaster) eventLoop(chRawLogs <-chan types.Log, chErr <-chan error)
 	var needsResubscribe bool
 	debounceResubscribe := time.NewTicker(1 * time.Second)
 	defer debounceResubscribe.Stop()
+	var latestSeenLogBlockNumber int64
 
 	b.logger.Debug("Starting the event loop")
 	for {
 		select {
 		case rawLog := <-chRawLogs:
-
 			b.logger.Debugw("Received a log",
 				"blockNumber", rawLog.BlockNumber, "blockHash", rawLog.BlockHash, "address", rawLog.Address)
-
 			b.onNewLog(rawLog)
+			latestSeenLogBlockNumber = int64(rawLog.BlockNumber)
```

Comment:
I think this will work, problem is that it could accidentally lead to a very deep backfill e.g. if a node is running jobs that only very rarely receive logs.

Reply:
This might be unavoidable to be honest. Can't think of another way to do it reliably.

Reply:
Yeah, I'm not sure how to get an accurate number from a non-racey source in that case without a major refactor (maybe even away from WS entirely). Would it make sense to put an upper bound on the range?

Reply:
There is no way to get a more accurate number. Probably makes sense to put an upper bound on the range, honestly thats what I thought `BLOCK_BACKFILL_DEPTH` was supposed to do. I'm a little confused as to what that actually does now.

Reply:
`BLOCK_BACKFILL_DEPTH` being an upper bound is actually the source of the problem here. The current behavior is to do `real-time-latest - BLOCK_BACKFILL_DEPTH`, so if the connection is down for more than `BLOCK_BACKFILL_DEPTH` we miss things. We would need to use a different bound for this to actually fix things.

Reply:
More context: this resub backfill logic differs from start-up logic, which depends on `highestSavedHead` from when the node stopped. So the node being down entirely for longer than `BLOCK_BACKFILL_DEPTH` is not a problem.

Reply:
If the logs are sparse the backfill would be deep, but it should be fast since the period between the last successful log and the beginning of the outage would have no logs to re process right?

Reply:
Yes, true. Processing would be minimal, although the fetching loop would still make `depth/batch` calls to the client and they may still have the same query performance even if no results are returned. Maybe something worth measuring.

Reply:
I think the node has to pull up each head to check the bloom filter, and then 1/7(?) false positives also require pulling up the logs themselves before seeing none are relevant.

Reply:
Eh not clear on false positives actually: https://github.com/ethereum/go-ethereum/issues/17613#issuecomment-431073923
Maybe only 0.5%?

Reply:
It's actually more complicated than this. During the downtime, we may have missed 'removed' logs that would have booted logs from the pool, so it's not safe to backfill from the latest seen log block number. I think we need to backfill from the _min block number in the pool_ (like an earlier draft) and also _clear the entire log pool_ as well. In the case of an empty pool, then we're back to choosing a sane default w/o much information.